src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerationsTests.cs

```diff
@@ -0,0 +1,1356 @@
+using System.Collections.Generic;
+
+public class Bar
+{
+    public void Sub(IEnumerable<int> j)
+    {
+        var i = Enumerable.Range(1, 10).ToArray();
+        foreach (var c in Enumerable.Range(1, 10))
+        {
+            i.Count();
+            [|j|].DefaultIfEmpty();
+        }
+    }
+}";
+            await VerifyCS.VerifyAnalyzerAsync(code);
+        }
+
+        [Fact]
+        public async Task TestInvocationsAfterForEachLoop()
+        {
+            var code = @"
+using System;
+using System.Linq;
+using System.Collections.Generic;
+
+public class Bar
+{
+    public void Sub()
+    {
+        IEnumerable<int> i = Enumerable.Range(1, 10);
+        foreach (var c in [|i|])
+        {
+        }
+        [|i|].Count();
+    }
+}";
+            await VerifyCS.VerifyAnalyzerAsync(code);
+        }
+
+        [Fact]
+        public async Task TestInvocationsAfterForEachLoopForArray()
+        {
+            var code = @"
+using System;
+using System.Linq;
+using System.Collections.Generic;
+
+public class Bar
+{
+    public void Sub()
+    {
+        var i = Enumerable.Range(1, 10).ToArray();
+        foreach (var c in i.Select(p => p))
+        {
+        }
+        i.Count();
+    }
+}";
+            await VerifyCS.VerifyAnalyzerAsync(code);
+        }
+
+        [Fact]
+        public async Task TestWhileLoop()
+        {
+            var code = @"
+using System;
+using System.Linq;
+using System.Collections.Generic;
+
+public class Bar
+{
+    public void Sub(IEnumerable<int> j)
+    {
+        IEnumerable<int> i = Enumerable.Range(1, 10);
+        while (true)
+        {
+            [|i|].ElementAt(100);
+            [|j|].ElementAtOrDefault(100);
+        }
+    }
+}";
+            await VerifyCS.VerifyAnalyzerAsync(code);
+        }
+
+        [Fact]
+        public async Task TestInvocationAfterIfBranchWithElse()
+        {
+            var code = @"
+using System;
+using System.Linq;
+using System.Collections.Generic;
+
+public class Bar
+{
+    public void Sub(int b, IEnumerable<int> j)
+    {
+        var i = Enumerable.Range(1, 10).ToArray();
+        if (b == 1)
+        {
+            i.First();
+            [|j|].FirstOrDefault();
+        }
+        else if (b == 3)
+        {
+            i.Last();
+            [|j|].LastOrDefault();
+        }
+        else
+        {
+            i.LongCount();
+            [|j|].Max();
+        }
+
+        i.Min();
+        [|j|].Single();
+    }
+}";
+
+            await VerifyCS.VerifyAnalyzerAsync(code);
+        }
+
+        [Fact]
+        public async Task TestInvocationAfterIfbranch()
+        {
+            var code = @"
+using System;
+using System.Linq;
+using System.Collections.Generic;
+
+public class Bar
+{
+    public void Sub(IEnumerable<int> h, int x)
+    {
+        IEnumerable<int> i = Enumerable.Range(1, 10);
+        if (x == 1)
+        {
+            [|i|].Except([|h|]).ToList();
+            [|i|].Intersect([|h|]).ToList();
+        }
+
+        i.Max();
```

Comment:
So, your analysis seems to be conservative in the sense that it only flags if enumerable is guaranteed to have multiple enumerations on all paths, but not if there is at least one path with multiple enumerations (which is the case for `i.Max();` here). That sounds reasonable, but even being more aggressive and flagging the latter case might be fine? Maybe you should talk to Mika and get her view.

Reply:
Can you please also add a test where `i.Max()` call is before the `if` statement?

Reply:
I will try to find her offline.

Reply:
Test added

Reply:
Synced with Mika, and she thinks we should flag `i.Max()`

Let me make code change here

Reply:
This is resolved in this commit https://github.com/dotnet/roslyn-analyzers/pull/5304/commits/c69dac0e61bdcf36a8294f3c5990cd71de7dba35